app/lib/actions/pre_upgrade_action.rb

```diff
@@ -0,0 +1,41 @@
+module Actions
+  module ForemanLeapp
+    class PreUpgradeAction < Actions::EntryAction
+      def self.subscribe
+        Actions::RemoteExecution::RunHostJob
+      end
+
+      def plan(job_invocation, host, template_invocation, *args)
+        return unless job_invocation.job_category == ::ForemanLeapp::JOB_CATEGORY
+        binding.pry
+        plan_self(trigger_success: triggering_action.output[:success],
+                  run_host_job_task_id: template_invocation.run_host_job_task_id,
+                  host_id: host.id)
+      end
+
+      def run
+        self.output.update(success: true) if self.input[:trigger_success]
+      end
+
+      def finalize(*args)
+        host = Host.find(self.input[:host_id])
+        task = ForemanTasks::Task.find(self.input[:run_host_job_task_id])
+        leapp_report = format_output(task.main_action.continuous_output.humanize)
+
+        PreupgradeReport.create_report(host, leapp_report)
+      end
+
+      private
+
+      def format_output(job_output)
+        lines = job_output.split(/\n/)
+        start_i = lines.index { |x| x.include?('===leap_upgrade_report_start===')}
+        end_i = lines.index { |x| x.include?('===leap_upgrade_report_end===')}
```

Comment:
How about doing this as a series of transformations instead of using indices? Something like

```ruby
job_outputs.each_line(chomp: true)
  .drop_while { |l| ! l.start_with? '===leap_upgrade_report_start===' }.drop(1)
  .take_while { |l| ! l.start_with? '===leap_upgrade_report_end===' }
  .reject(&:empty?)
  .join('')
```